developer-documentation.md

```diff
@@ -79,3 +79,332 @@ address of any Google account that you own, such as an `@gmail.com` account.
 - Executing `npm start` or `npm test` automatically starts the Datastore Emulator and shuts it down afterwards.
+
+## Adding a new API
+
+This section outlines the steps to consider when adding a new API.
+
+Note: For all new APIs, please consider using [OpenAPI](https://www.openapis.org/).
+With OpenAPI, developers can write a specification for their API and have code
+generated for them on both the frontend and backend. This helps remove the
+burden of manually writing data models and data encoding and decoding for both sides.
+There is a tool installed as a devDependency called
+[openapi-generator-cli](https://github.com/OpenAPITools/openapi-generator-cli)
+to do the generation of the code.
+
+The specification follows OpenAPI version 3 and is located at [api.yaml](./api.yaml).
+
+Below are steps to help guide a developer along with a relatable example that follows the same steps.
+
+### Step 0: Additional Tools To Help
+
+If using Visual Studio Code, install the following extensions. (These are pre-installed if using the devcontainer)
+- [OpenAPI (Swagger) Editor](https://marketplace.visualstudio.com/items?itemName=42Crunch.vscode-openapi)
+
+### Step 1: Add the path and operations
+
+*Before completing this step, read the [Paths and Operations](https://swagger.io/docs/specification/paths-and-operations/) and [Describing Parameters](https://swagger.io/docs/specification/describing-parameters/) OpenAPI docs*
+
+#### Step 1a: Add the Path
+
+1. Open the spec file and navigate to the top level `paths` object.
+2. Add the desired path. All paths already have the `/api/v0` prefix (check the `servers` object). Do not include that prefix when adding the path.
+
+<details>
+  <summary>Example (click to expand)</summary>
+  
+  #### api.yaml
+  ```yaml
+  paths:
+    /features/{feature_id}:
+  ```
+</details>
+
+
+
+#### Step 1b: Add Operations
+
+Operations = HTTP verbs. (e.g. GET, POST, PUT, etc)
+
+- Add the operation(s) under the path.
+- Ensure each operation has a `summary`, `description` and `operationId`
+- If your path has path parameters, describe the parameters now too.
+  - Mark these are required as well.
+
+<details>
+  <summary>Example (click to expand)</summary>
+  
+  #### api.yaml
+  ```yaml
+  paths:
+    /features/{feature_id}:
+      get:
+        summary: Get a feature by ID.
+        description: >-
+          Get a feature by ID. More details about this here.
+          Also, can do more comments
+        operationId: getFeatureById
+        parameters:
+          - name: feature_id
+            in: path
+            description: Feature ID
+            required: true
+            schema:
```

Comment:
What do you think about using the YAML "flow" syntax to compress some of the most common lines?
https://yaml.org/spec/1.2.2/#74-flow-collection-styles
E.g., `schema: {"type": "integer", "format": "int64"}`

Reply:
I would be hesitant to do this syntax for a few reasons:
- I would want to be consistent with every object.
- Current tooling doesn't work with this. (Lose auto-completed suggestions)

In the meantime, I remember there was a question about splitting this into multiple files. I wasn't sure at the time but I did some digging and found out that it can indeed be [done](https://swagger.io/docs/specification/using-ref/). (see `$ref Syntax`) If length becomes a problem, we can explore this.

Reply:
I figured out how to split the files out. I have updated the instructions to do it. The other PRs have been updated to the new modular system.

Another good resource about is https://blog.pchudzik.com/202004/open-api-and-external-ref/